cmd/vspd/spentticket.go

```diff
@@ -0,0 +1,164 @@
+
+	if startHeight > endHeight {
+		return nil, 0, fmt.Errorf("start height %d greater than best block height %d",
+			startHeight, endHeight)
+	}
+
+	numBlocks := int(1 + endHeight - startHeight)
+
+	// Only log if checking a larger number of blocks to avoid spam.
+	if numBlocks > 5 {
+		v.log.Debugf("Scanning %d blocks for %s",
+			numBlocks, pluralize(len(toCheck), "spent ticket"))
+	}
+
+	// Get commitment address payment script for each ticket.
+	type ticketTuple struct {
+		dbTicket database.Ticket
+		pkScript []byte
+	}
+
+	tickets := make(map[string]ticketTuple)
+	for _, ticket := range toCheck {
+		parsedAddr, err := stdaddr.DecodeAddress(ticket.CommitmentAddress, params)
+		if err != nil {
+			return nil, 0, err
+		}
+		_, script := parsedAddr.PaymentScript()
+
+		tickets[ticket.Hash] = ticketTuple{
+			dbTicket: ticket,
+			pkScript: script,
+		}
+	}
+
+	spent := make(map[string]spentTicket)
+
+	for iHeight := startHeight; iHeight <= endHeight; iHeight++ {
+		iHash, err := dcrdClient.GetBlockHash(iHeight)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		iBlock, err := dcrdClient.GetBlockVerbose(iHash)
+		if err != nil {
+			return nil, 0, fmt.Errorf("error retrieving block %s: %w", iHash, err)
+		}
+
+		key, filter, err := dcrdClient.GetCFilterV2(iBlock.Hash, iBlock.MerkleRoot)
+		if err != nil {
+			return nil, 0, err
+		}
+
+	outer:
+		for ticketHash, ticket := range tickets {
+			if filter.Match(key, ticket.pkScript) {
+				// Ticket is likely spent in block. Confirm.
+				for _, blkTx := range append(iBlock.RawTx, iBlock.RawSTx...) {
+					if !txSpendsOutput(blkTx, ticketHash) {
+						continue
+					}
+
+					// Confirmed - ticket is spent in block.
+
+					decoded, err := hex.DecodeString(blkTx.Hex)
+					if err != nil {
+						return nil, 0, fmt.Errorf("could not decode hex %q: %w", blkTx.Hex, err)
+					}
+
+					msgTx := &wire.MsgTx{}
+					err = msgTx.FromBytes(decoded)
+					if err != nil {
+						return nil, 0, fmt.Errorf("could not get msgTx from hex %q: %w", blkTx.Hex, err)
+					}
+
+					spent[ticketHash] = spentTicket{
+						dbTicket:     ticket.dbTicket,
+						expiryHeight: ticket.dbTicket.PurchaseHeight + int64(params.TicketMaturity) + int64(params.TicketExpiry),
+						heightSpent:  iBlock.Height,
+						spendingTx:   msgTx,
+					}
+
+					// Remove this ticket and continue with the next one.
+					delete(tickets, ticketHash)
+					continue outer
+				}
+
+				// Ticket is not spent in block.
+			}
+		}
+
+		if len(tickets) == 0 {
+			// Found spenders for all tickets, stop searching.
+			break
+		}
+	}
+
+	return spent, endHeight, nil
+}
+
+// txSpendsOutput returns true if the passed tx has an input that spends the
+// specified output.
+func txSpendsOutput(tx types.TxRawResult, outputHash string) bool {
+	for i := range tx.Vin {
+		input := &tx.Vin[i]
+		if input.Vout == 0 && input.Txid == outputHash {
+			return true // Found spender.
+		}
+	}
+	return false
+}
+
+func pluralize(n int, s string) string {
```

Comment:
This works for the only usage there currently is, but I'd suggest stealing [pickNoun](https://github.com/decred/dcrd/blob/0a35adc02120805f5185cc0a53eb2674711e3013/log.go#L179-L186) for this purpose because not all words can be made plural by just adding an s.

Reply:
I'm going to leave this as is because I personally prefer the simplicity at the call sites. It will be enough to update later if required. Even in dcrd which uses pickNoun a whole bunch I notice that it never actually does anything more than add an "s" at the moment.

I've added a comment to make the limitation really explicit.